### PR TITLE
Save widget

### PR DESCRIPTION
**Pull request by andresgnlez**
Pivotal task: https://www.pivotaltracker.com/story/show/141225215

Modal buttons distribution are pending on design (https://www.pivotaltracker.com/story/show/143021909)

NOTES:

- `print` function depends on the embed widget (the plan is to take a screenshot of this)
- `download` function will be developed later because of complexity.
-  Inside `share`, we will need to provide a real URL for share the widget and embed one.

These points are writte down in Pivotal: https://www.pivotaltracker.com/story/show/143164293

### CODE REVIEW

File: app/assets/javascripts/components/data_portal/modals/modal-save-widget.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      title: 'Save widget',
+      // See App.Component.Modal for details about this option
+      showTitle: true,
+      // See App.Component.Modal for details about this option
+      allowScroll: true,
+      // See App.Component.Modal for details about this option
+      isAbsolute: true,
+      // See App.Component.Modal for details about this option
+      footer: '<button type="button" class="c-button -medium -white -outline js-button">Add to report</button><div class="group-button"><button type="button" class="c-button -medium -white -disabled js-print">Print</button><button type="button" class="c-button -medium -white -disabled">Download</button><button type="button" data-slide-index="0" class="c-button -medium -white js-slide-button">Share</button></div>',
+      // TO-DO: add description
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/modals/modal-save-widget.js]: - [ ] add description

File: app/assets/javascripts/components/shared/modal.js [hunk 2]
```diff
@@ -24,6 +24,11 @@
         showTitle: false,
         // Title of the modal – mandatory for accessibility
         title: 'Untitled modal',
+        // Allows the modal container have scroll
+        allowScroll: false,
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/shared/modal.js]: Why did you add that? We didn't need it before and you're not using it for the save modal.

File: app/assets/javascripts/components/data_portal/modals/modal-save-widget.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    // TO-DO: take a screenshot of the embed
+    _onPrint: function () {
+      console.warn('WIP!')
+    },
+
+    _renderWidget: function () {
+      var widgetConfig = _.extend({}, this.options.widgetConfig, {
+        el: $(this.el).find('.js-widget-container'),
+        showToolbar: this.defaults.widgetConfig.showToolbar
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/modals/modal-save-widget.js]: This line is useless, you already extends with `this.options.widgetConfig`.

File: app/assets/javascripts/components/data_portal/slide.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+(function (App) {
+  App.Component.Slide = Backbone.View.extend({
+
+    defaults: {},
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/slide.js]: Please add the description of `callback`. Something like:
```javascript
defaults: {
  // Callback to be executed when the slide closes
  callback: function () {}
}
```

File: app/assets/javascripts/components/data_portal/slide.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+      // Outgoing slide styles
+      if (this.el.classList.contains('-enter')) {
+        this.el.classList.remove('-enter');
+        this.el.classList.add('-leave');
+      }
+    },
+
+    onCallback: function () {
+      if (this.options.callback && typeof this.options.callback === 'function') {
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/slide.js]: Because we're adding a default value for `this.options.callback` in `defaults`, you don't need to check anything, just call it.

File: app/assets/javascripts/views/data_portal/slides/ShareView.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    className: 'slide',
+
+    events: {
+      'click .js-back': '_onBack',
+      'click .js-copy': '_onCopy',
+      'blur  .js-copy': '_onBlur'
+    },
+
+    initialize: function (settings) {
```
Comment (clementprdhomme) [inline app/assets/javascripts/views/data_portal/slides/ShareView.js]: Useless. If you don't do anything here, just remove the declaration of `initialize`.

File: app/assets/javascripts/views/data_portal/slides/ShareView.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      this.constructor.__super__.initialize.call(this, settings);
+    },
+
+    _onBack: function () {
```
Comment (clementprdhomme) [inline app/assets/javascripts/views/data_portal/slides/ShareView.js]: This is useless too. Instead of calling `_onBack`, call directly `this.onCallback`.

File: app/assets/javascripts/views/data_portal/slides/ShareView.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    _onCopy: function (e) {
+      var copyButton = e.currentTarget;
+
+      window.getSelection().removeAllRanges();
+      var rangeSelection = document.createRange();
+      rangeSelection.selectNode(copyButton.previousElementSibling);
+      window.getSelection().addRange(rangeSelection)
+
+      var confirmation = document.execCommand('copy');
```
Comment (clementprdhomme) [inline app/assets/javascripts/views/data_portal/slides/ShareView.js]: Missing `try {} catch(e) {}`. Not all the browsers support `document.execCommand`.

File: app/assets/javascripts/components/data_portal/slide.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        this.options.callback();
+      }
+    },
+
+    render: function () {
+      this.el.innerHTML = this.content;
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/slide.js]: Let's move `this.content` to `this.options.content` to make it clean.

File: app/assets/stylesheets/_base.scss [hunk 10]
```diff
@@ -61,6 +61,13 @@ select {
   width: 100%;
   height: 100%;
   transform: translate(-50%, -50%);
+
+  &.-absolute {
```
Comment (clementprdhomme) [inline app/assets/stylesheets/_base.scss]: I wouldn't add this rules to the global scope but only to the modal. We don't use it anywhere else so let's keep it local.

File: app/assets/stylesheets/components/data-portal/c-slider.scss [hunk 11]
```diff
@@ -0,0 +1,50 @@
+
+.c-slider {
+  position: absolute;
+  top: 0;
+  left: 0;
+  width: 100%;
+  height: 100%;
+  transform: translate(calc(100% + 1px), 0);
```
Comment (clementprdhomme) [inline app/assets/stylesheets/components/data-portal/c-slider.scss]: Is there any reason for `100% + 1px`?

File: app/assets/stylesheets/components/data-portal/c-slider.scss [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    position: absolute;
+    width: 100%;
+    height: 100%;
+    padding: 30px;
+
+    transition: .15s ease-out transform;
+    background-color: $color-4;
+
+    &.-enter {
+      transform: translate(calc(-100% - 1px), 0);
```
Comment (clementprdhomme) [inline app/assets/stylesheets/components/data-portal/c-slider.scss]: Same here: `-100% - 1px`?

File: app/assets/stylesheets/components/shared/c-button.scss [hunk 13]
```diff
@@ -13,6 +13,10 @@
 
   cursor: pointer;
 
+  &.-disabled {
```
Comment (clementprdhomme) [inline app/assets/stylesheets/components/shared/c-button.scss]: Maybe you could also add the styles for `[disabled]`.

File: app/assets/javascripts/components/data_portal/modals/modal-save-widget.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+      // See App.Component.Modal for details about this option
+      title: 'Save widget',
+      // See App.Component.Modal for details about this option
+      showTitle: true,
+      // See App.Component.Modal for details about this option
+      allowScroll: true,
+      // See App.Component.Modal for details about this option
+      isAbsolute: true,
+      // See App.Component.Modal for details about this option
+      footer: '<button type="button" class="c-button -medium -white -outline js-button">Add to report</button><div class="group-button"><button type="button" class="c-button -medium -white -disabled js-print">Print</button><button type="button" class="c-button -medium -white -disabled">Download</button><button type="button" data-slide-index="0" class="c-button -medium -white js-slide-button">Share</button></div>',
```
Comment (clementprdhomme) [inline app/assets/javascripts/components/data_portal/modals/modal-save-widget.js]: The styles of the buttons are wrong: the first one should be white and the other ones with just the outline.

**Comment by clementprdhomme** [comment]
Regarding the issue of the scroll jump with the auto-focus on, here is the fix: don't execute `this.focusableElements[0].focus();` in `modal.js` before the modal is positioned.

The easiest solution which maintains the modularity is to define a method called `afterRender`
 in `modal.js` and to execute it at the end of `render`. The content of the method should be the three lines you've commented.

Then, in `App.Component.ModalSaveWidget`, create the method `afterRender` with the following content:
```javascript
afterRender: function () {
  this._setPosition();
  this._renderWidget();
  this._renderSlides();
  this.constructor.__super__.afterRender.apply(this);
}
```

Make sure to remove the three lines from `render`.

We also have another issue: by default, one of the element of the slides will be focused. To prevent that, let's make something of the widget focusable. For example, add `tabindex="0"` to `.c-widget` in `modal-save-widget.jst.ejs`.

Finally, as I've told you before, you'll need to change those three lines in `modal.js`:
```javascript
// We add some attributes for the accessibility
this.el.setAttribute('role', 'dialog');
this.el.setAttribute('aria-label', this.options.title);
this.el.setAttribute('tabindex', '0');
```
so the attributes are added to `js-modal-content`.

File: app/assets/javascripts/components/data_portal/modals/modal-save-widget.js (hunk 3 above)
Comment (andresgnlez) [inline app/assets/javascripts/components/data_portal/modals/modal-save-widget.js]: That's what I thought, but for some reason, I was inheriting parent's value and it wasn't overriding with child's one. That's why I had to set it manually. Anyway, I'll take a look again.

File: app/assets/javascripts/components/shared/modal.js (hunk 2 above)
Comment (andresgnlez) [inline app/assets/javascripts/components/shared/modal.js]: We are: this is the way to allow scroll the page despite of having the modal opened. But this doesn't happen in the rest of modals, you can not scroll once the modal is opened.

File: app/assets/javascripts/views/data_portal/slides/ShareView.js (hunk 8 above)
Comment (andresgnlez) [inline app/assets/javascripts/views/data_portal/slides/ShareView.js]: They do: http://caniuse.com/#search=document.execCommand. I checked it before implement it.

File: app/assets/stylesheets/components/data-portal/c-slider.scss (hunk 11 above)
Comment (andresgnlez) [inline app/assets/stylesheets/components/data-portal/c-slider.scss]: For some reason, when I moved the slide to show/hide it, there was always 1px of the slider showing up. This fixes it.

File: app/assets/stylesheets/components/data-portal/c-slider.scss (hunk 12 above)
Comment (andresgnlez) [inline app/assets/stylesheets/components/data-portal/c-slider.scss]: https://github.com/Vizzuality/i2i/pull/94#discussion_r110097312

File: app/assets/stylesheets/components/shared/c-button.scss (hunk 13 above)
Comment (andresgnlez) [inline app/assets/stylesheets/components/shared/c-button.scss]: Agree.

**Comment by andresgnlez** [comment]
Great! I'll try your solution in https://github.com/Vizzuality/i2i/pull/94#issuecomment-291932340 👍

File: app/assets/stylesheets/_base.scss (hunk 10 above)
Comment (andresgnlez) [inline app/assets/stylesheets/_base.scss]: This style affects the wrapper that contains the modal and veil themselves, it's not something specific of the modal.

File: app/assets/javascripts/components/shared/modal.js (hunk 2 above)
Comment (clementprdhomme) [inline app/assets/javascripts/components/shared/modal.js]: By default, the scroll is always available so there's no reason to "allow" it. It already is.

Also, the rest of the modals didn't have any issue. We're just letting the user scrolls if they want. Moreover, disabling the scroll as you're doing makes mobile devices scroll to top, so it's more an inconvenient than a solution.

File: app/assets/javascripts/views/data_portal/slides/ShareView.js (hunk 8 above)
Comment (clementprdhomme) [inline app/assets/javascripts/views/data_portal/slides/ShareView.js]: Ok, you're right!

Nevertheless, caniuse is not entirely true: the support for copy [has been added in Safari 10](https://developer.apple.com/library/content/releasenotes/General/WhatsNewInSafari/Articles/Safari_10_0.html) and wasn't available before. You already check for the value, so all good.

File: app/assets/stylesheets/_base.scss (hunk 10 above)
Comment (clementprdhomme) [inline app/assets/stylesheets/_base.scss]: My fault! Didn't see the style was within `.modal-container`.

File: app/assets/stylesheets/components/data-portal/c-slider.scss (hunk 11 above)
Comment (clementprdhomme) [inline app/assets/stylesheets/components/data-portal/c-slider.scss]: Might be a half pixel then.

File: app/assets/javascripts/components/shared/modal.js (hunk 2 above)
Comment (andresgnlez) [inline app/assets/javascripts/components/shared/modal.js]: It's not. That's why we have `this.body.classList.toggle('_no-scroll', !this.options.allowScroll);` (previously it was just `this.body.classList.add('_no-scroll')`)

Comment (clementprdhomme) [inline app/assets/javascripts/components/shared/modal.js]: The scroll is blocked on Chrome and FF but not on Safari. This was my confusion.
